components/service/glean/src/main/java/mozilla/components/service/glean/scheduler/MetricsPingScheduler.kt

```diff
@@ -0,0 +1,85 @@
+import android.content.SharedPreferences
+import mozilla.components.support.base.log.logger.Logger
+import java.util.concurrent.TimeUnit
+
+/**
+ * MetricsPingScheduler facilitates sending and scheduling of metrics pings
+ */
+internal class MetricsPingScheduler(val applicationContext: Context) {
+    private val logger = Logger("glean/MetricsPingScheduler")
+    internal val sharedPreferences: SharedPreferences by lazy { getMetricsPrefs() }
+
+    companion object {
+        const val LAST_METRICS_PING_TIMESTAMP_KEY = "last_metrics_ping_timestamp"
+        const val STORE_NAME = "metrics"
+        const val PING_INTERVAL_HOURS = 24L
+    }
+
+    /**
+     * Check to see if a metrics ping can be sent.
+     *
+     * This will compare the current system time to the last ping time stored in [SharedPreferences]
+     * to determine whether [PING_INTERVAL_HOURS] has elapsed.
+     *
+     * @return true if the time interval has elapsed or if this is the first attempt to send a
+     *         metrics ping, otherwise false.
+     */
+    fun canSendPing(): Boolean {
+        var lastPingTime = 0L
+
+        try {
+            lastPingTime = sharedPreferences.getLong(LAST_METRICS_PING_TIMESTAMP_KEY, 0L)
+        } catch (e: ClassCastException) {
+            // Intentionally doing nothing here, if LAST_METRICS_PING_TIMESTAMP_KEY's value is
+            // not a Long, then we will let lastPingTime remain 0L for the purposes of comparison
+            logger.error("MetricsPingScheduler last ping time stored was not a Long value")
+        }
+
+        val currentTime = System.currentTimeMillis()
+
+        if (lastPingTime == 0L ||
+            currentTime - lastPingTime >= TimeUnit.HOURS.toMillis(PING_INTERVAL_HOURS)) {
+            return true
+        }
+
+        return false
+    }
+
+    /**
+     * Update the persisted timestamp when the metrics ping is sent.
+     *
+     * This is called after sending a metrics ping in order to timestamp when the last ping was
+     * sent in order to maintain the proper interval between pings. See [canSendPing].
+     *
+     * @param timestamp The timestamp to store, which defaults to [System.currentTimeMillis].
+     * @return Returns the time that was stored in [SharedPreferences] for comparision.
+     */
+    fun updateSentTimestamp(timestamp: Long = System.currentTimeMillis()): Long {
+        sharedPreferences.edit()
+            .putLong(LAST_METRICS_PING_TIMESTAMP_KEY, timestamp)
+            .apply()
+
+        return timestamp
+    }
+
+    /**
+     * Initializes the [SharedPreferences] where persisted metrics ping information is stored.
+     */
+    private fun getMetricsPrefs(): SharedPreferences {
+        return applicationContext.getSharedPreferences(this.javaClass.simpleName, Context.MODE_PRIVATE)
+    }
+
+    /**
+     * Test only function to clear stored data
+     */
+    internal fun clearSchedulerData() {
+        sharedPreferences?.edit().clear().apply()
```

Comment:
Why is it assuming a nullable here? Should we do the same at the other callsites?